File: src/core/gui/Table.ts
import { Widget, WidgetOptions } from "./Widget";
import { bind_children_to, el } from "./dom";
import { ListProperty } from "../observable/property/list/ListProperty";
import { Disposer } from "../observable/Disposer";
import "./Table.css";
import Logger = require("js-logger");
import { Disposable } from "../observable/Disposable";

const logger = Logger.get("core/gui/Table");

export type Column<T> = {
    key?: string;
    title: string;
    fixed?: boolean;
    width: number;
    input?: boolean;
    text_align?: string;
    tooltip?: (value: T) => string;
    sortable?: boolean;
    render_cell(value: T, disposer: Disposer): string | HTMLElement;
    footer?: {
        render_cell(): string;
        tooltip?(): string;
    };
};

export enum SortDirection {
    Asc,
    Desc,
}

export type TableOptions<T> = WidgetOptions & {
    values: ListProperty<T>;
    columns: Column<T>[];
    sort?(sort_columns: { column: Column<T>; direction: SortDirection }[]): void;
};

export class Table<T> extends Widget {
    readonly element = el.table({ class: "core_Table" });

    private readonly tbody_element = el.tbody();
    private readonly footer_row_element?: HTMLTableRowElement;
    private readonly values: ListProperty<T>;
    private readonly columns: Column<T>[];

    constructor(options: TableOptions<T>) {
        super(options);

        this.values = options.values;
        this.columns = options.columns;

        let sort_columns: { column: Column<T>; direction: SortDirection }[] = [];

        const thead_element = el.thead();
        const header_tr_element = el.tr();

        let left = 0;
        let has_footer = false;

        header_tr_element.append(
            ...this.columns.map((column, index) => {
                const th = el.th(
                    { data: { index: index.toString() } },
                    el.span({ text: column.title }),
                );

                if (column.fixed) {
                    th.style.position = "sticky";
                    th.style.left = `${left}px`;
                    left += column.width;
                }

                th.style.width = `${column.width}px`;

                if (column.footer) {
                    has_footer = true;
                }

                return th;
            }),
        );

        const sort = options.sort;

        if (sort) {
            header_tr_element.onmousedown = e => {
                if (e.target instanceof HTMLElement) {
                    let element: HTMLElement = e.target;

                    for (let i = 0; i < 5; i++) {
                        if (element.dataset.index) {
                            break;
                        } else if (element.parentElement) {
                            element = element.parentElement;
                        } else {
                            return;
                        }
                    }

                    if (!element.dataset.index) return;

                    const index = parseInt(element.dataset.index, 10);
                    const column = this.columns[index];
                    if (!column.sortable) return;

                    const existing_index = sort_columns.findIndex(sc => sc.column === column);

                    if (existing_index === 0) {
                        const sc = sort_columns[0];
                        sc.direction =
                            sc.direction === SortDirection.Asc
                                ? SortDirection.Desc
                                : SortDirection.Asc;
                    } else {
                        if (existing_index !== -1) {
                            sort_columns.splice(existing_index, 1);
                        }

                        sort_columns.unshift({ column, direction: SortDirection.Asc });
                    }

                    sort(sort_columns);
                }
            };
        }

        thead_element.append(header_tr_element);
        this.tbody_element = el.tbody();
        this.element.append(thead_element, this.tbody_element);

        if (has_footer) {
            this.footer_row_element = el.tr();
            this.element.append(el.tfoot({}, this.footer_row_element));
            this.create_footer();
        }

        this.disposables(
            bind_children_to(this.tbody_element, this.values, this.create_row),
            this.values.observe(this.update_footer),
        );

        this.finalize_construction(Table.prototype);
    }

    private create_row = (value: T, index: number): [HTMLTableRowElement, Disposable] => {
        const disposer = new Disposer();
        let left = 0;

        return [
            el.tr(
                {},
                ...this.columns.map((column, i) => {
                    const cell = column.fixed ? el.th() : el.td();

                    try {
                        const content = column.render_cell(value, disposer);

                        cell.append(content);

                        if (column.input) cell.classList.add("input");

                        if (column.fixed) {
                            cell.classList.add("fixed");
                            cell.style.left = `${left}px`;
                            left += column.width || 0;
                        }

                        cell.style.width = `${column.width}px`;

                        if (column.text_align) cell.style.textAlign = column.text_align;

                        if (column.tooltip) cell.title = column.tooltip(value);
                    } catch (e) {
                        logger.warn(
                            `Error while rendering cell for index ${index}, column ${i}.`,
                            e,
                        );
                    }

                    return cell;
                }),
            ),
            disposer,
        ];
    };

    private create_footer(): void {
        const footer_cells: HTMLTableHeaderCellElement[] = [];
        let left = 0;

        for (let i = 0; i < this.columns.length; i++) {
            const column = this.columns[i];
            const cell = el.th();

            cell.style.width = `${column.width}px`;

            if (column.fixed) {
                cell.classList.add("fixed");
                cell.style.left = `${left}px`;
                left += column.width || 0;
            }

            if (column.footer) {
                cell.textContent = column.footer.render_cell();
                cell.title = column.footer.tooltip ? column.footer.tooltip() : "";
            }

            if (column.text_align) cell.style.textAlign = column.text_align;

            footer_cells.push(cell);
        }

        this.footer_row_element!.append(...footer_cells);
    }

    private update_footer = (): void => {
        if (!this.footer_row_element) return;

        const col_count = this.columns.length;

        for (let i = 0; i < col_count; i++) {
            const column = this.columns[i];

            if (column.footer) {
                const cell = this.footer_row_element.children[i] as HTMLTableHeaderCellElement;
                cell.textContent = column.footer.render_cell();
                cell.title = column.footer.tooltip ? column.footer.tooltip() : "";
            }
        }
    };
}

File: webui/src/main/kotlin/world/phantasmal/webui/widgets/Table.kt
package world.phantasmal.webui.widgets

import org.w3c.dom.*
import world.phantasmal.core.disposable.Disposable
import world.phantasmal.core.disposable.Disposer
import world.phantasmal.observable.cell.Cell
import world.phantasmal.observable.cell.observeNow
import world.phantasmal.observable.cell.trueCell
import world.phantasmal.webui.LoadingStatus
import world.phantasmal.webui.controllers.Column
import world.phantasmal.webui.controllers.TableController
import world.phantasmal.webui.dom.*

class Table<T>(
    visible: Cell<Boolean> = trueCell(),
    enabled: Cell<Boolean> = trueCell(),
    private val ctrl: TableController<T>,
    private val className: String? = null,
    /**
     * Can return a [Widget].
     */
    private val renderCell: (T, Column<T>) -> Any,
) : Widget(visible, enabled) {
    override fun Node.createElement() =
        table {
            className = "pw-table"

            this@Table.className?.let { classList.add(it) }

            ctrl.loadingStatus?.let { loadingStatus ->
                div {
                    className = "pw-table-notification"

                    observeNow(loadingStatus) { status ->
                        when (status) {
                            LoadingStatus.Uninitialized,
                            LoadingStatus.InitialLoad,
                            -> {
                                hidden = false
                                innerText = "Loading..."
                            }
                            LoadingStatus.Error -> {
                                hidden = false
                                innerText = "An error occurred while loading this table."
                            }
                            else -> {
                                hidden = true
                            }
                        }
                    }
                }
            }

            thead {
                tr {
                    className = "pw-table-row pw-table-header-row"

                    addDisposable(bindChildrenTo(
                        this,
                        ctrl.columns,
                        createChild = { column, _ ->
                            createHeaderRowCell(column)
                        },
                        after = {
                            positionFixedColumns(row = this, headerRow = true)
                        },
                    ))
                }
            }
            tbody {
                bindDisposableChildrenTo(ctrl.values) { value, _ ->
                    val rowDisposer = Disposer()

                    val row = tr {
                        className = "pw-table-row"

                        addDisposable(bindChildrenTo(
                            this,
                            ctrl.columns,
                            createChild = { column, _ ->
                                createRowCell(column, value, rowDisposer)
                            },
                            after = {
                                positionFixedColumns(row = this, headerRow = false)
                            },
                        ))
                    }

                    Pair(row, rowDisposer)
                }
            }

            if (ctrl.hasFooter) {
                tfoot {
                    tr {
                        className = "pw-table-row pw-table-footer-row"

                        addDisposable(bindDisposableChildrenTo(
                            this,
                            ctrl.columns,
                            createChild = { column, _ ->
                                createFooterRowCell(column)
                            },
                            after = {
                                positionFixedColumns(row = this, headerRow = false)
                            },
                        ))
                    }
                }
            }
        }

    private fun Node.createHeaderRowCell(column: Column<T>): HTMLTableCellElement =
        th {
            className = "pw-table-cell pw-table-header-cell"

            column.headerClassName?.let { classList.add(it) }

            textContent = column.title

            style.width = "${column.width}px"

            if (column.sortable) {
                onmousedown = { e ->
                    if (e.buttons.toInt() == 1) {
                        ctrl.sortByColumn(column)
                    }
                }
            }
        }

    private fun Node.createRowCell(
        column: Column<T>,
        value: T,
        rowDisposer: Disposer,
    ): HTMLTableCellElement =
        td {
            className = "pw-table-cell pw-table-body-cell"

            column.className?.let { classList.add(it) }

            val child = renderCell(value, column)

            if (child is Widget) {
                rowDisposer.add(child)
                append(child.element)
            } else {
                append(child)
            }

            if (column.input) {
                classList.add("pw-table-cell-input")
            }

            style.width = "${column.width}px"
            column.tooltip(value)?.let { title = it }
            column.textAlign?.let { style.textAlign = it }
        }

    private fun Node.createFooterRowCell(
        column: Column<T>,
    ): Pair<HTMLTableCellElement, Disposable> {
        val disposer = Disposer()

        val cell = th {
            className = "pw-table-cell pw-table-footer-cell"

            style.width = "${column.width}px"
            column.textAlign?.let { style.textAlign = it }

            disposer.add(column.footer.observeNow { textContent = it ?: "" })
            disposer.add(column.footerTooltip.observeNow { title = it ?: "" })
        }

        return Pair(cell, disposer)
    }

    private fun positionFixedColumns(row: HTMLTableRowElement, headerRow: Boolean) {
        val columns = ctrl.columns.value
        var left = 0

        for (index in columns.indices) {
            val el = row.children[index].unsafeCast<HTMLElement>()

            if (index < ctrl.fixedColumns) {
                el.style.position = "sticky"
                el.style.left = "${left}px"

                if (!headerRow) {
                    el.classList.add("pw-table-cell-fixed")
                }
            } else {
                el.style.position = ""
            }

            left += columns[index].width
        }
    }

    companion object {
        init {
            @Suppress("CssUnresolvedCustomProperty", "CssUnusedSymbol")
            // language=css
            style("""
                .pw-table {
                    position: relative;
                    display: block;
                    box-sizing: border-box;
                    overflow: auto;
                    background-color: var(--pw-bg-color);
                    border-collapse: collapse;
                }

                .pw-table-notification {
                    position: absolute;
                    top: 0;
                    bottom: 0;
                    left: 0;
                    right: 0;
                    display: grid;
                    grid-template: 100% / 100%;
                    place-items: center;
                    text-align: center;
                    color: var(--pw-text-color-disabled);
                    font-size: 20px;
                }

                .pw-table > thead {
                    position: sticky;
                    display: inline-block;
                    top: 0;
                    z-index: 2;
                }

                .pw-table > tbody, .pw-table > tfoot {
                    user-select: text;
                    cursor: text;
                }

                .pw-table-row {
                    display: flex;
                    align-items: stretch;
                }

                .pw-table-header-row {
                    position: sticky;
                    top: 0;
                }

                .pw-table-header-cell {
                    font-weight: bold;
                    display: flex;
                    flex-direction: column;
                    justify-content: center;
                    overflow: hidden;
                }

                .pw-table-footer-cell {
                    font-weight: bold;
                    text-align: left;
                }

                .pw-table-cell {
                    box-sizing: border-box;
                    overflow: hidden;
                    text-overflow: ellipsis;
                    padding: 3px 6px;
                    border-right: var(--pw-border);
                    border-bottom: var(--pw-border);
                    background-color: var(--pw-bg-color);
                }

                .pw-table-body-cell {
                    white-space: nowrap;
                }

                .pw-table-body-cell {
                    text-align: left;
                }

                .pw-table-cell-fixed {
                    font-weight: bold;
                    position: sticky;
                    text-align: left;
                }

                .pw-table-cell-input {
                    padding: 0;
                    overflow: visible;
                }

                .pw-table-cell-input > .pw-input {
                    z-index: 0;
                    height: 100%;
                    width: 100%;
                    border: none;
                }

                .pw-table-cell-input > .pw-input:hover,
                .pw-table-cell-input > .pw-input:focus-within {
                    margin: -1px;
                    height: calc(100% + 2px);
                    width: calc(100% + 2px);
                }

                .pw-table-cell-input > .pw-input:hover {
                    z-index: 4;
                    border: var(--pw-input-border-hover);
                }

                .pw-table-cell-input > .pw-input:focus-within {
                    z-index: 6;
                    border: var(--pw-input-border-focus);
                }
            """.trimIndent())
        }
    }
}
